[ADD] DNS hoster INWX

Kariton opened the pull request:
Add [INWX ](https://inwx.de) by utilizing [inwx.collection](https://github.com/inwx/ansible-collection)

* [ADD] DNS provider 'INWX' to argument_specs.yml

* [ADD] INWX sanity checks

* [ADD] INWX create

* [ADD] INWX cleanup

Kariton committed:
[ADD] DNS hoster INWX (#1)

* [ADD] DNS provider 'INWX' to argument_specs.yml

* [ADD] INWX sanity checks

* [ADD] INWX create

* [ADD] INWX cleanup

Kariton commented:
INWX enforces a TTL of min. 300 seconds.
This is respected while creation  - and waiting for propagation.

I did not include this in the sanity checks - as there arent any for the other provider.

Kariton committed:
[FIX] typo in argument_specs.yml

Kariton commented:
ohhh. and i fu**** up...
my PR `dns-inwx-create.yml` includes: 
```yml
- name: Wait for DNS entries to propagate
  community.dns.wait_for_txt:
    always_ask_default_resolver: false
```` 

will remove `always_ask_default_resolver: false`!

Kariton committed:
[FIX] typo 'wait_for_txt' dns-inwx-create.yml

Remove `always_ask_default_resolver: false` from module `community.dns.wait_for_txt` while 'Wait for DNS entries to propagate'

felixfontein commented on roles/acme_certificate/tasks/dns-inwx-create.yml:
```diff
@@ -0,0 +1,25 @@
+    domain: "{{ item.key | community.dns.get_registrable_domain }}"
+    type: TXT
+    record: "{{ item.key | community.dns.remove_registrable_domain }}"
+    value: "{{ item.value | first }}"
```
Hmm, what about the other values in the list? (This happens when you want a cert for `foo.example.com` and `*.foo.example.com`: then you have two TXT entries for `_acme-challenge.foo.example.com`.)

Kariton replied:
Okay. Thanks.
I totally missed that. Unfortunately this does not work then.
I need to loop that task for each `item.value` - atm it will create the literal list as TXT record.
```
# dig -t TXT @9.9.9.9 +short _acme-challenge.foo.example.com.
"['XXX', 'YYY']"
```

To do so i need to think about a proper way - any idea?
My first plan is to set a custom fact to do so. (data manipulation is not my favorite)
OR they (INWX INC.) would add the ability to loop the list within the module - its their own.

AND `solo: true` must get removed then - it ensures that a record does not exist multiple times.

Kariton replied:
WORKS:
`acme_certificate_domains: "['*.foo.example.com', 'foo.example.com']"`

```yml
- name: Creating challenge DNS entries for {{ ', '.join(acme_certificate_domains + acme_certificate_ips) }} via INWX
  inwx.collection.dns:
    state: present
    domain: "{{ item.0.key | community.dns.get_registrable_domain }}"
    type: TXT
    record: "{{ item.0.key | community.dns.remove_registrable_domain }}"
    value: "{{ item.1 }}"
    ttl: 300
    username: "{{ acme_certificate_inwx_username | default(omit) }}"
    password: "{{ acme_certificate_inwx_password | default(omit) }}"
  delegate_to: localhost
  run_once: true
  loop: "{{ acme_certificate_INTERNAL_challenge.challenge_data_dns | dict2items | subelements('value') }}"
```

Kariton replied:
There is an open request for a """problem""" like this. 
(handle `value` input if it is a list)

https://github.com/inwx/ansible-collection/issues/9

Kariton replied:
Would you implement this:
```yml
- name: Creating challenge DNS entries for {{ ', '.join(acme_certificate_domains + acme_certificate_ips) }} via INWX
  inwx.collection.dns:
    state: present
    domain: "{{ item.0.key | community.dns.get_registrable_domain }}"
    type: TXT
    record: "{{ item.0.key | community.dns.remove_registrable_domain }}"
    value: "{{ item.1 }}"
    solo: "{{ 'true' if ansible_loop.first else 'false' }}"
    ttl: 300
    username: "{{ acme_certificate_inwx_username | default(omit) }}"
    password: "{{ acme_certificate_inwx_password | default(omit) }}"
  delegate_to: localhost
  run_once: true
  loop_control:
    extended: yes  
  loop: "{{ acme_certificate_INTERNAL_challenge.challenge_data_dns | dict2items | subelements('value') }}"
```
(this does remove all "same" DNS records if it creates the first one)

or see what the feature request on their side would do? (mentioned above)

Kariton replied:
https://github.com/inwx/ansible-collection/issues/9#issuecomment-951357893
> Hey, as there seems to be some interest in this feature I'm going to implement this. It will probably be released with new record types (https://github.com/inwx/ansible-collection/pull/12) in 1.3.0. But I can't guarantee any specific release date.

felixfontein replied:
Cool! I guess for now we can use the loop with `subelements` so that we have something that works, and once a new version if out bump the requirements and use that new feature.

felixfontein commented on roles/acme_certificate/tasks/dns-inwx-sanity.yml:
```diff
@@ -0,0 +1,8 @@
+  ansible.builtin.assert:
+    that:
+      - acme_certificate_inwx_username is defined
+      - acme_certificate_inwx_password is defined
```
I think we also need to check whether the inwx.collection collection is installed, if we don't want to add it as a requirement to galaxy.yml (which I would prefer not to). Unfortunately that's not possible with Ansible right now, so we have to stick to documentation for now :)

(I'll try to figure out how to properly test for this, probably it will need a new action or lookup plugin or something like that.)

Kariton replied:
I wouldnt add this as requirement to the `galaxy.yml` ether.
An update for the docs is IMO better. :)
Will provide one after this task does work as intended.

felixfontein reviewed (commented):
Thanks for your contribution! Can you please add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/development_process.html#changelogs-how-to) (`minor_changes`)?

You also need to update docs/docsite/rst/general_role_parameters.rst and docs/docsite/rst/acme_certificate_role.rst.

Kariton committed:
[FIX] INWX and docs (#2)

* [FIX] multiple record values dns-inwx-create.yml
this is necessary as long as https://github.com/inwx/ansible-collection/issues/9 is not resolved.

* Update general_role_parameters.rst

* Update acme_certificate_role.rst

Kariton committed:
Update dns-inwx-cleanup.yml (#3)

Kariton commented:
QA is important. NOW it does what it should do.

Kariton committed:
I should update the general requirements

Kariton commented:
and learn how to keep GIT clean...

Thanks for your understanding.
And a relay big thanks for your lovely role! :)

Kariton committed:
Create 19-add-inwx-dns-provider.yaml (#4)

Kariton committed:
[FIX] Sanity (Ⓐstable-2.10) / Extra Sanity

`ERROR: docs/docsite/rst/general_role_parameters.rst:95:0: Bullet list ends without a blank line; unexpected unindent.` /
`./docs/docsite/rst/general_role_parameters.rst:95:0: (WARNING/2) Bullet list ends without a blank line; unexpected unindent.`

I hope that is the needed fix.

Kariton commented:
@felixfontein 
How do I resolve this - can I see where the problem is?
```
Run yamllint .
  yamllint .
  shell: /usr/bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/3.8.12/x64
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/3.8.12/x64/lib
Error: [trailing-spaces] trailing spaces
Error: Process completed with exit code 1.
```

felixfontein commented on docs/docsite/rst/general_role_parameters.rst:
```diff
@@ -87,12 +87,13 @@ DNS Challenges
 If DNS challenges are used, the following variables define how the challenges can be fulfilled:
 
-- ``acme_certificate_dns_provider``: must be one of ``route53``, ``hosttech``, and ``ns1``. Each needs more information:
+- ``acme_certificate_dns_provider``: must be one of ``route53``, ``hosttech``, ``ns1`` and ``inwx``. Each needs more information:
```
```suggestion
- ``acme_certificate_dns_provider``: must be one of ``route53``, ``hosttech``, ``ns1``, and ``inwx``. Each needs more information:
```
I really like the Oxford comma :)

felixfontein commented on docs/docsite/rst/general_role_parameters.rst:
```diff
@@ -105,3 +106,10 @@ For ``ns1`` (`ns1.com <https://ns1.com>`_) the external ``ns1_record`` module ne
+
+For ``inwx`` (`inwx.de <https://inwx.de>`_) the available Ansible Galaxy collection ``inwx.collection`` (`galaxy.ansible.com <https://galaxy.ansible.com/inwx/collection>`_) needs to be installed.
+  
+It's as simple as: ``ansible-galaxy collection install inwx.collection``
```
```suggestion
It is as simple as: ``ansible-galaxy collection install inwx.collection``
```

felixfontein commented on docs/docsite/rst/requirements.rst:
```diff
@@ -7,4 +7,4 @@ The collection requires the Python `cryptography <https://pypi.org/project/crypt
 The ``openssl`` binary must also be available in the executable path on the controller. It is needed by the :ref:`community.crypto.acme_certificate <ansible_collections.community.crypto.acme_certificate_module>` module in case ``cryptography`` is not installed, and it is used for certificate chain validation.
 
-If DNS challenges are used, there can be other requirements depending on the DNS provider. For example, for Amazon's Route 53, the Ansible :ref:`community.aws.route53 <ansible_collections.community.aws.route53_module>` requires the Python `boto <https://pypi.org/project/boto/>`_ package. If Hosttech DNS challenges are used, the `lxml <https://pypi.org/project/lxml/>`_ package needs to be installe ...
+If DNS challenges are used, there can be other requirements depending on the DNS provider. For example, for Amazon's Route 53, the Ansible :ref:`community.aws.route53 <ansible_collections.community.aws.route53_module>` requires the Python `boto <https://pypi.org/project/boto/>`_ package. If Hosttech DNS challenges are used, the `lxml <https://pypi.org/project/lxml/>`_ package needs to be installe ...
```
```suggestion
If DNS challenges are used, there can be other requirements depending on the DNS provider. For example, for Amazon's Route 53, the Ansible :ref:`community.aws.route53 <ansible_collections.community.aws.route53_module>` requires the Python `boto <https://pypi.org/project/boto/>`_ package. If Hosttech DNS challenges are used, the `lxml <https://pypi.org/project/lxml/>`_ package needs to be installed. If DNS challenges with NS1 or INWX are used, the corresponding modules respectively collections must be installed. See below for more information.
```

felixfontein commented on roles/acme_certificate/meta/argument_specs.yml:
```diff
@@ -119,6 +120,20 @@ argument_specs:
+          - When using I(acme_certificate_dns_provider=inwx), the INWX account password must be passed in this option.
+          - Please note that INWX support depends on the collection C(inwx.collection). Please check out the
+            documentation on how to install it.
+          - Please note that the collection C(inwx.collection) does currently not support 2FA (two factor authentication)
+            in order to use this module you need to disable it.
```
```suggestion
          - Please note that the collection C(inwx.collection) does currently not support 2FA (two factor authentication).
            In order to use this module you need to disable it.
```

felixfontein commented on roles/acme_certificate/tasks/dns-inwx-create.yml:
```diff
@@ -0,0 +1,27 @@
+  delegate_to: localhost
+  run_once: true
+  loop_control:
+    extended: yes  
```
```suggestion
    extended: yes
```
Here are the trailing spaces.

I ran `yamllint` manually, it was then much more helpful with its error message:
```
$ yamllint .
./roles/acme_certificate/tasks/dns-inwx-create.yml
  18:18     error    trailing spaces  (trailing-spaces)
```
No idea why in CI it is so ... unhelpful :)

felixfontein commented on docs/docsite/rst/general_role_parameters.rst:
```diff
@@ -105,3 +106,10 @@ For ``ns1`` (`ns1.com <https://ns1.com>`_) the external ``ns1_record`` module ne
 Once NS1 converts their `set of modules <https://github.com/ns1/ns1-ansible-modules>`_ into a `collection <https://docs.ansible.com/ansible/latest/dev_guide/developing_collections.html>`_, it will become a lot easier to install and use them (`NS1 tracking issue <https://github.com/ns1/ns1-ansible-modules/issues/32>`_).
+
+Setting up INWX modules
+~~~~~~~~~~~~~~~~~~~~~~
```
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~
```

Kariton committed:
Update docs/docsite/rst/general_role_parameters.rst

Co-authored-by: Felix Fontein <felix@fontein.de>

Kariton committed:
Update docs/docsite/rst/general_role_parameters.rst

Co-authored-by: Felix Fontein <felix@fontein.de>

Kariton committed:
Update docs/docsite/rst/requirements.rst

Co-authored-by: Felix Fontein <felix@fontein.de>

Kariton committed:
Update roles/acme_certificate/meta/argument_specs.yml

Co-authored-by: Felix Fontein <felix@fontein.de>

Kariton committed:
Update roles/acme_certificate/tasks/dns-inwx-create.yml

Co-authored-by: Felix Fontein <felix@fontein.de>

Kariton committed:
Update docs/docsite/rst/general_role_parameters.rst

Co-authored-by: Felix Fontein <felix@fontein.de>

felixfontein commented on docs/docsite/rst/general_role_parameters.rst:
```diff
@@ -87,12 +87,13 @@ DNS Challenges
+  - For ``inwx`` (`inwx.de <https://inwx.de>`_) the credentials have to be passed as ``acme_certificate_inwx_username`` and ``acme_certificate_inwx_password``. Please keep in mind that 2FA (two factor authentication) is currently not supported and needs do be disabled. Related Issue: (`inwx/ansible-collection#3 <https://github.com/inwx/ansible-collection/issues/3>`_)
 
-Please note that the DNS challenge code is not perfect. The Route 53, Hosttech and NS1 functionality has been tested.
+Please note that the DNS challenge code is not perfect. The Route 53, Hosttech, NS1 and INWX functionality has been tested.
```
```suggestion
Please note that the DNS challenge code is not perfect. The Route 53, Hosttech, NS1, and INWX functionality has been tested.
```

felixfontein commented on roles/acme_certificate/tasks/dns-inwx-create.yml:
```diff
@@ -0,0 +1,27 @@
+    type: TXT
+    record: "{{ item.0.key | community.dns.remove_registrable_domain }}"
+    value: "{{ item.1 }}"
+    solo: "{{ 'true' if ansible_loop.first else 'false' }}"
```
This can be simplified:
```suggestion
    solo: "{{ ansible_loop.first }}"
```

felixfontein commented on changelogs/fragments/19-add-inwx-dns-provider.yaml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+- "The role ``felixfontein.acme.acme_certificate`` can now handle the DNS provider INWX."
```
```suggestion
- "The role ``felixfontein.acme.acme_certificate`` can now handle the DNS provider INWX (https://github.com/felixfontein/ansible-acme/pull/19)."
```

felixfontein reviewed (commented):
Besides this, LGTM!

Kariton committed:
Update docs/docsite/rst/general_role_parameters.rst

Co-authored-by: Felix Fontein <felix@fontein.de>

Kariton committed:
Update roles/acme_certificate/tasks/dns-inwx-create.yml

Co-authored-by: Felix Fontein <felix@fontein.de>

Kariton committed:
Update changelogs/fragments/19-add-inwx-dns-provider.yaml

Co-authored-by: Felix Fontein <felix@fontein.de>

felixfontein commented:
@Kariton thanks a lot for contributing this!

I'll probably create a new release later this week or next weekend.
